File: labs/laba2/laba2_java/ArrClass.java
class ArrClass {
    private int dim;
    private int threadNum;
    private int threadCount = 0;
    private Thread[] threads;
    private int[] arr;
    private Object lockerForMin = new Object();
    private Object lockerForCount = new Object();
    private long min = Long.MAX_VALUE;
    private int minIndex = -1;

    public ArrClass(int dim, int threadNum) {
        this.dim = dim;
        this.threadNum = threadNum;
        this.threads = new Thread[threadNum];
        this.arr = new int[dim];
    }

    public void initArr() {
        for (int i = 0; i < dim; i++) {
            arr[i] = i;
        }
        arr[90000000] = -500;
    }

    public MinResult partMin(int startIndex, int finishIndex) {
        long localMin = Long.MAX_VALUE;
        int localIndex = -1;
        
        for (int i = startIndex; i < finishIndex; i++) {
            if (arr[i] < localMin) {
                localMin = arr[i];
                localIndex = i;
            }
        }
        return new MinResult(localMin, localIndex);
    }

    public MinResult parallelMin() {
        int bound = dim/threadNum;
        int startBound = 0;

        for (int i = 0; i < threadNum - 1; i++) {
            final int start = startBound;
            final int finish = startBound + bound;
            threads[i] = new Thread(() -> starterThread(new Bound(start, finish)));
            threads[i].start();
            startBound += bound;
        }
        
        final int lastStart = startBound;
        threads[threadNum - 1] = new Thread(() -> starterThread(new Bound(lastStart, dim)));
        threads[threadNum - 1].start();

        synchronized (lockerForCount) {
            while (threadCount < threadNum) {
                try {
                    lockerForCount.wait();
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                }
            }
        }
        
        return new MinResult(this.min, this.minIndex);
    }

    private void starterThread(Bound bound) {
        MinResult res = partMin(bound.startIndex, bound.finishIndex);

        synchronized (lockerForMin) {
            collectMin(res.min, res.index);
        }
        incThreadCount();
    }

    private void incThreadCount() {
        synchronized (lockerForCount) {
            threadCount++;
            lockerForCount.notify();
        }
    }

    private void collectMin(long localMin, int localIndex) {
        if (localMin < this.min) {
            this.min = localMin;
            this.minIndex = localIndex;
        }
    }
}

File: labs/laba2/laba2_sharp/ArrClass.cs
namespace ThreadMinSharp;
public class ArrClass
{
    private readonly int dim;
        private readonly int threadNum;
        private static int threadCount = 0;
        private  readonly Thread[] thread;
        private  readonly int[] arr;   
        private  readonly object lockerForMin = new object();
        private  readonly object lockerForCount = new object();
        public long Min { get; private set; } = long.MaxValue;
        public int MinIndex { get; private set; } = -1;

        public ArrClass(int dim,int threadNum)
        {
            this.dim = dim;
            this.threadNum  = threadNum;
            this.thread = new Thread[threadNum];
            this.arr = new int[dim];
        }
        public long ParallelMin()
        {
            long bound = dim/threadNum;
            long startBound = 0;
            for(int i = 0; i < threadNum-1; i++)
            {
                thread[i] = new Thread(StarterThread);
                thread[i].Start(new Bound(startBound, startBound+bound));
                startBound+=bound;
            }
            thread[threadNum-1] = new Thread(StarterThread);
            thread[threadNum-1].Start(new Bound(startBound, dim));

            lock (lockerForCount)
            {
                while (threadCount < threadNum)
                {
                    Monitor.Wait(lockerForCount);
                }
            }
            return Min;
        }

        public void InitArr()
        {
            for (int i = 0; i < dim; i++)
            {
                arr[i] = i;
            }

            arr[100000] = -500;
        }
        
        private void StarterThread(object param)
        {
            if (param is Bound)
            {
                var res = PartMin((param as Bound).StartIndex, (param as Bound).FinishIndex);

                lock (lockerForMin)
                {
                    CollectMin(res.min, res.index);
                }
                IncThreadCount();
            }
        }

        private  void IncThreadCount()
        {
            lock (lockerForCount)
            {
                threadCount++;
                Monitor.Pulse(lockerForCount);
            }
        }
       
        public void CollectMin(long min,int index)
        {
            if (min < this.Min)
            {
                this.Min=min;
                this.MinIndex=index;
            }
        }

        public (long min, int index) PartMin(long startIndex, long finishIndex)
        {
            long min = long.MaxValue;
            int index = -1;
            for (long i = startIndex; i < finishIndex; i++)
            {
                if (arr[i]<min){
                min=arr[i];
                index=(int)i;
                }
            }
            return (min,index);
        }
    }